stream.go

```diff
@@ -0,0 +1,510 @@
+package molecule
+
+import (
+	"bytes"
+	"io"
+	"math"
+
+	"google.golang.org/protobuf/encoding/protowire"
+)
+
+const (
+	// wire types; see https://developers.google.com/protocol-buffers/docs/encoding
+	wtVarint          int = 0
+	wt64Bit           int = 1
+	wtLengthDelimited int = 2
+	wt32Bit           int = 5
+
+	// defaultBufferSize is the default size for buffers used for embedded
+	// values, which must first be written to a buffer to determine their
+	// length.  This is not used if BufferFactory is set.
+	defaultBufferSize int = 1024 * 16
+)
+
+// A ProtoStream supports writing protobuf data in a streaming fashion.  Its methods
+// will write their output to the wrapped `io.Writer`.  Zero values are not included.
+//
+// ProtoStream instances are *not* threadsafe and *not* re-entrant.
+type ProtoStream struct {
+	// outputWriter is the writer to which the protobuf-encoded bytes are written
+	outputWriter io.Writer
+
+	// scratchBuffer is a buffer used and re-used for generating output.  Each method
+	// should begin by resetting this buffer.
+	scratchBuffer []byte
+
+	// scratchArray is a second, very small array used for packed encodings.  It is large
+	// enough to fit two max-size varints (10 bytes each) without reallocation
+	scratchArray [20]byte
+
+	// childStream is a ProtoStream used to implement `Embedded`, and reused for
+	// multiple calls.
+	childStream *ProtoStream
+
+	// childBuffer is the buffer to which `childStream` writes.
+	childBuffer *bytes.Buffer
+
+	// BufferFactory creates new, empty buffers as needed.  Users may override
+	// this function to provide pre-initialized buffers of a larger size, or
+	// from a buffer pool, for example.
+	BufferFactory func() []byte
+}
+
+// NewProtoStream creates a new ProtoStream.  This ProtoStream _cannot be used_ until
+// Reset is called.
+func NewProtoStream() *ProtoStream {
+	return &ProtoStream{
+		scratchBuffer: make([]byte, 0, defaultBufferSize),
+		childStream:   nil,
+		childBuffer:   nil,
+		BufferFactory: func() []byte { return make([]byte, 0, defaultBufferSize) },
+	}
+}
+
+// Reset sets the Writer to which this ProtoStream streams.  This must be
+// called before any other methods, and can be called again to change the
+// output Writer.
+func (ps *ProtoStream) Reset(outputWriter io.Writer) {
+	ps.outputWriter = outputWriter
```

Comment:
maybe:

```golang
p.scratchBuffer = p.scratchBuffer[:0]
for i := range p.scratchArray {
    p.scratchArray[I] = 0
}
```

to prevent any minor bugs from leaking information between resets

Reply:
Hm, why reset `scratchArray` and not `scratchBuffer`?

I get the desire for safety here, but zeroing out buffers is potentially expensive, so I'm not sure it should be a built-in feature.

I dislike that Go doesn't statically protect against reading allocated-but-uninitialized memory.  Or perhaps a different way to look at that is that it does not allow recategorizing memory as uninitialized and thus protected against unintended re-use.

Reply:
```golang
p.scratchBuffer = p.scratchBuffer[:0]
```

is free, so lets do that at least? It literally just changes the length of the slice header to 0, no expensive zeroing

I think memclearing the fixed size scratchArray won't be too costly either but I'm ok skipping on that one since its way less likely to contain sensitive data

Reply:
Sounds good!